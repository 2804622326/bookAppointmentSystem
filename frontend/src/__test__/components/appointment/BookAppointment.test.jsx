import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import BookAppointment from '../../../components/appointment/BookAppointment.jsx';

jest.mock('react-router-dom', () => ({
  useParams: () => ({ recipientId: '2' }),
}));

jest.mock('react-datepicker', () => (props) => (
  <input data-testid={props['data-testid'] || 'date'} onChange={(e) => props.onChange(e.target.value)} placeholder={props.placeholderText} />
));

jest.mock('../../../components/pet/PetEntry.jsx', () => () => <div data-testid='pet-entry' />);

jest.mock('../../../components/utils/utilities', () => ({
  dateTimeFormatter: jest.fn(() => ({ formattedDate: '2024-01-01', formattedTime: '12:00' })),
}));

jest.mock('../../../components/appointment/AppointmentService.js', () => ({
  bookAppointment: jest.fn(),
}));
import { bookAppointment } from '../../../components/appointment/AppointmentService.js';
bookAppointment.mockResolvedValue({ message: 'Booked' });

jest.mock('../../../components/hooks/UseMessageAlerts', () => ({ __esModule: true, default: jest.fn() }));
import useAlerts from '../../../components/hooks/UseMessageAlerts';
const alerts = {
  setSuccessMessage: jest.fn(),
  setShowSuccessAlert: jest.fn(),
  setErrorMessage: jest.fn(),
  setShowErrorAlert: jest.fn(),
};
useAlerts.mockReturnValue({
  successMessage: '',
  showSuccessAlert: false,
  showErrorAlert: false,
  ...alerts,
});

test('submits form and shows success', async () => {
  localStorage.setItem('userId', '1');
  const { asFragment } = render(<BookAppointment />);

  fireEvent.change(screen.getByPlaceholderText('Choose a date'), { target: { value: '2024-05-10' } });
  fireEvent.change(screen.getByPlaceholderText('Select time'), { target: { value: '10:00' } });
  fireEvent.change(screen.getAllByRole('textbox')[2], { target: { value: 'Checkup' } });

  fireEvent.click(screen.getByRole('button', { name: /book appointment/i }));
  await waitFor(() => expect(bookAppointment).toHaveBeenCalled());
  expect(bookAppointment).toHaveBeenCalledWith('1', '2', expect.any(Object));
  expect(alerts.setSuccessMessage).toHaveBeenCalledWith('Booked');
  expect(alerts.setShowSuccessAlert).toHaveBeenCalledWith(true);
  expect(asFragment()).toMatchSnapshot();
});
