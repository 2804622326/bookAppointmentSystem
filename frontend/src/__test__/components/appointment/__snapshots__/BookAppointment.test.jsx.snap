// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`submits form and shows success 1`] = `
<DocumentFragment>
  <div
    class="mt-5 container"
  >
    <div
      class="justify-content-center row"
    >
      <div
        class="col-lg-6 col-md-10 col-sm-12"
      >
        <form
          class=""
        >
          <div
            class="shadow mb-5 card"
          >
            <h5
              class="text-center card-header"
            >
               Appointment Booking Form
            </h5>
            <div
              class="card-body"
            >
              <fieldset
                class="field-set mb-4"
              >
                <legend
                  class="text-center"
                >
                  Appointment Date and Time
                </legend>
                <div
                  class="mb-4 row"
                >
                  <div
                    class="col"
                  >
                    <input
                      data-testid="date"
                      placeholder="Choose a date"
                    />
                  </div>
                  <div
                    class="col"
                  >
                    <input
                      data-testid="date"
                      placeholder="Select time"
                    />
                  </div>
                </div>
              </fieldset>
              <div
                class="mb-4"
              >
                <label
                  class="form-label"
                >
                  Reason for appointment
                </label>
                <textarea
                  class="form-control"
                  name="reason"
                  required=""
                  rows="3"
                />
              </div>
              <h5
                class="text-center"
              >
                Appointment Pet Information
              </h5>
              <div
                data-testid="pet-entry"
              />
              <div
                class="d-flex justify-content-center mb-3"
              >
                <button
                  class="me-2 btn btn-primary btn-sm"
                  type="button"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                </button>
                <button
                  class="me-2 btn btn-outline-primary btn-sm"
                  type="submit"
                >
                    Book Appointment
                </button>
                <button
                  class="btn btn-outline-info btn-sm"
                  type="button"
                >
                  Reset
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
